### PR TITLE
OpenConceptLab/ocl_issues#147: Users can view collections that are marked private if the user query is specified

### PR DESCRIPTION
**Pull request by Karuhanga**
Address https://github.com/OpenConceptLab/ocl_issues/issues/147

### CODE REVIEW

**Comment by Karuhanga** [comment]
Still adding permissions tests.

**Comment by rkorytkowski** [comment]
For permission tests please see https://github.com/OpenConceptLab/oclapi/tree/master/integration-tests

File: integration-tests/src/collection.spec.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            privateCollectionOwnedByUser1,
+            org1,
+            user3NotInOrg1,
+            user2InOrg1,
+            user1InOrg1,
+        ];
+        for(let i in items) await del(items[i].url, adminToken);
+    });
+
+    test('user can retrieve own collections', async () => {
```
Comment (Karuhanga) [inline integration-tests/src/collection.spec.ts]: This test fails. It shouldn't.

File: ocl/oclapi/filters.py [hunk 2]
```diff
@@ -138,4 +144,27 @@ def _filter_queryset(self, request, queryset, view, sqs):
...
     def filter_queryset(self, request, queryset, view):
-        return self._filter_queryset(request, queryset, view, SearchQuerySet())
\ No newline at end of file
+        return self._filter_queryset(request, queryset, view, RelatedSearchQuerySet())
+
+
+class ConceptContainerPermissionedSearchFilter(HaystackSearchFilter):
+    def filter_queryset(self, request, queryset, view):
+        current_user = request.user
+        print queryset.query, queryset, 'original qs \n\n'
```
Comment (Karuhanga) [inline ocl/oclapi/filters.py]:
```
SELECT collection_collection.id, collection_collection.public_access, collection_collection.created_at, collection_collection.updated_at, collection_collection.created_by, collection_collection.updated_by, collection_collection.is_active, collection_collection.extras, collection_collection.uri, collection_collection.mnemonic, collection_collection.parent_type_id, collection_collection.parent_id, collection_collection.name, collection_collection.full_name, collection_collection.default_locale, collection_collection.supported_locales, collection_collection.website, collection_collection.description, collection_collection.external_id, collection_collection.custom_validation_schema, collection_collection.references, collection_collection.collection_type, collection_collection.preferred_source, collection_collection.repository_type, collection_collection.custom_resources_linked_source FROM collection_collection WHERE (collection_collection.is_active = True  AND collection_collection.parent_id = 5d667a70394d010e84a51efc  AND collection_collection.parent_type_id = 5d664271394d01003b4a8133 ) [<Collection: 25166>, <Collection: 73767>] original
```

File: ocl/oclapi/filters.py [hunk 3]
```diff
@@ -138,4 +144,27 @@ def _filter_queryset(self, request, queryset, view, sqs):
...
+
+        permissioned_qs = queryset.exclude(public_access=ACCESS_TYPE_NONE)
+        print permissioned_qs.query, permissioned_qs, 'qs from public access \n\n'
```
Comment (Karuhanga) [inline ocl/oclapi/filters.py]:
```
SELECT collection_collection.id, collection_collection.public_access, collection_collection.created_at, collection_collection.updated_at, collection_collection.created_by, collection_collection.updated_by, collection_collection.is_active, collection_collection.extras, collection_collection.uri, collection_collection.mnemonic, collection_collection.parent_type_id, collection_collection.parent_id, collection_collection.name, collection_collection.full_name, collection_collection.default_locale, collection_collection.supported_locales, collection_collection.website, collection_collection.description, collection_collection.external_id, collection_collection.custom_validation_schema, collection_collection.references, collection_collection.collection_type, collection_collection.preferred_source, collection_collection.repository_type, collection_collection.custom_resources_linked_source FROM collection_collection WHERE (collection_collection.is_active = True  AND collection_collection.parent_id = 5d667a70394d010e84a51efc  AND collection_collection.parent_type_id = 5d664271394d01003b4a8133  AND NOT (collection_collection.public_access = None )) [<Collection: 73767>] from public access
```

File: ocl/oclapi/filters.py [hunk 4]
```diff
@@ -138,4 +144,27 @@ def _filter_queryset(self, request, queryset, view, sqs):
...
+
+        if not current_user.is_anonymous():
+            user_profile = UserProfile.objects.get(user=current_user)
+            private_collections = queryset.filter(
+                Q(parent_id=user_profile.id, parent_type=ContentType.objects.get_for_model(UserProfile)) |
+                Q(parent_id__in=user_profile.organizations, parent_type=ContentType.objects.get_for_model(Organization))
+            )
+            print private_collections.query, private_collections, 'qs from permissions \n\n'
```
Comment (Karuhanga) [inline ocl/oclapi/filters.py]:
```
SELECT collection_collection.id, collection_collection.public_access, collection_collection.created_at, collection_collection.updated_at, collection_collection.created_by, collection_collection.updated_by, collection_collection.is_active, collection_collection.extras, collection_collection.uri, collection_collection.mnemonic, collection_collection.parent_type_id, collection_collection.parent_id, collection_collection.name, collection_collection.full_name, collection_collection.default_locale, collection_collection.supported_locales, collection_collection.website, collection_collection.description, collection_collection.external_id, collection_collection.custom_validation_schema, collection_collection.references, collection_collection.collection_type, collection_collection.preferred_source, collection_collection.repository_type, collection_collection.custom_resources_linked_source FROM collection_collection WHERE (collection_collection.is_active = True  AND collection_collection.parent_id = 5d667a70394d010e84a51efc  AND collection_collection.parent_type_id = 5d664271394d01003b4a8133  AND ((collection_collection.parent_id = 5d667a70394d010e84a51efc  AND collection_collection.parent_type_id = 5d664271394d01003b4a8133 ) OR (collection_collection.parent_type_id = 5d664271394d01003b4a8131  AND collection_collection.parent_id IN (5d667a70394d010e84a51f01)))) [<Collection: 25166>, <Collection: 73767>] queryset from permissions
```

File: ocl/oclapi/filters.py [hunk 5]
```diff
@@ -138,4 +144,27 @@ def _filter_queryset(self, request, queryset, view, sqs):
...
+
+            permissioned_qs |= private_collections
+
+            print permissioned_qs.query, permissioned_qs, 'qs from both', '\n'
```
Comment (Karuhanga) [inline ocl/oclapi/filters.py]:
```
SELECT collection_collection.id, collection_collection.public_access, collection_collection.created_at, collection_collection.updated_at, collection_collection.created_by, collection_collection.updated_by, collection_collection.is_active, collection_collection.extras, collection_collection.uri, collection_collection.mnemonic, collection_collection.parent_type_id, collection_collection.parent_id, collection_collection.name, collection_collection.full_name, collection_collection.default_locale, collection_collection.supported_locales, collection_collection.website, collection_collection.description, collection_collection.external_id, collection_collection.custom_validation_schema, collection_collection.references, collection_collection.collection_type, collection_collection.preferred_source, collection_collection.repository_type, collection_collection.custom_resources_linked_source FROM collection_collection WHERE ((collection_collection.is_active = True  AND collection_collection.parent_id = 5d667a70394d010e84a51efc  AND collection_collection.parent_type_id = 5d664271394d01003b4a8133  AND NOT (collection_collection.public_access = None )) OR (collection_collection.is_active = True  AND collection_collection.parent_id = 5d667a70394d010e84a51efc  AND collection_collection.parent_type_id = 5d664271394d01003b4a8133  AND ((collection_collection.parent_id = 5d667a70394d010e84a51efc  AND collection_collection.parent_type_id = 5d664271394d01003b4a8133 ) OR (collection_collection.parent_type_id = 5d664271394d01003b4a8131  AND collection_collection.parent_id IN (5d667a70394d010e84a51f01))))) [<Collection: 25166>, <Collection: 71445>, <Collection: 73767>, <Collection: 29655>] both
```

File: ocl/oclapi/filters.py [hunk 6]
```diff
@@ -138,4 +144,27 @@ def _filter_queryset(self, request, queryset, view, sqs):
 
 class HaystackSearchFilter(BaseHaystackSearchFilter):
     def filter_queryset(self, request, queryset, view):
-        return self._filter_queryset(request, queryset, view, SearchQuerySet())
\ No newline at end of file
+        return self._filter_queryset(request, queryset, view, RelatedSearchQuerySet())
+
+
+class ConceptContainerPermissionedSearchFilter(HaystackSearchFilter):
+    def filter_queryset(self, request, queryset, view):
```
Comment (Karuhanga) [inline ocl/oclapi/filters.py]: There seems to be some incorrect behavior with with performing an `OR` on a queryset. Doing an or seems to return more results than are in either query sets. I suspect this is coming from `django-no-rel` but I cannot be sure until I investigate farther.

**Comment by Karuhanga** [comment]
@rkorytkowski 
This is a problem I have been able to trace down to `mongodb-engine` by comparing the queryset query and what is actually executed on mongo. I'd really appreciate help with this.
```
original qs

SELECT collection_collection.id, collection_collection.public_access, collection_collection.created_at, collection_collection.updated_at, collection_collection.created_by, collection_collection.updated_by, collection_collection.is_active, collection_collection.extras, collection_collection.uri, collection_collection.mnemonic, collection_collection.parent_type_id, collection_collection.parent_id, collection_collection.name, collection_collection.full_name, collection_collection.default_locale, collection_collection.supported_locales, collection_collection.website, collection_collection.description, collection_collection.external_id, collection_collection.custom_validation_schema, collection_collection.references, collection_collection.collection_type, collection_collection.preferred_source, collection_collection.repository_type, collection_collection.custom_resources_linked_source FROM collection_collection WHERE (collection_collection.is_active = True  AND collection_collection.parent_id = 5d721432394d010a2a73c0e3  AND collection_collection.parent_type_id = 5d664271394d01003b4a8133 ) [<Collection: 86285>, <Collection: 86772>]
{
        "op" : "query",
        "ns" : "ocl.collection_collection",
        "query" : {
                "find" : "collection_collection",
                "filter" : {
                        "parent_id" : "5d721432394d010a2a73c0e3",
                        "is_active" : true,
                        "parent_type_id" : ObjectId("5d664271394d01003b4a8133")
                },
                "ntoreturn" : 21
        },
        "ts" : ISODate("2019-09-06T08:09:23.162Z")
}


qs from public access 
SELECT collection_collection.id, collection_collection.public_access, collection_collection.created_at, collection_collection.updated_at, collection_collection.created_by, collection_collection.updated_by, collection_collection.is_active, collection_collection.extras, collection_collection.uri, collection_collection.mnemonic, collection_collection.parent_type_id, collection_collection.parent_id, collection_collection.name, collection_collection.full_name, collection_collection.default_locale, collection_collection.supported_locales, collection_collection.website, collection_collection.description, collection_collection.external_id, collection_collection.custom_validation_schema, collection_collection.references, collection_collection.collection_type, collection_collection.preferred_source, collection_collection.repository_type, collection_collection.custom_resources_linked_source FROM collection_collection WHERE (collection_collection.is_active = True  AND collection_collection.parent_id = 5d721432394d010a2a73c0e3  AND collection_collection.parent_type_id = 5d664271394d01003b4a8133  AND NOT (collection_collection.public_access = None )) [<Collection: 86772>]
{
        "op" : "query",
        "ns" : "ocl.collection_collection",
        "query" : {
                "find" : "collection_collection",
                "filter" : {
                        "parent_id" : "5d721432394d010a2a73c0e3",
                        "public_access" : {
                                "$ne" : "None"
                        },
                        "is_active" : true,
                        "parent_type_id" : ObjectId("5d664271394d01003b4a8133")
                },
                "ntoreturn" : 21
        },
        "ts" : ISODate("2019-09-06T08:09:23.166Z")
}


qs from permissions
SELECT collection_collection.id, collection_collection.public_access, collection_collection.created_at, collection_collection.updated_at, collection_collection.created_by, collection_collection.updated_by, collection_collection.is_active, collection_collection.extras, collection_collection.uri, collection_collection.mnemonic, collection_collection.parent_type_id, collection_collection.parent_id, collection_collection.name, collection_collection.full_name, collection_collection.default_locale, collection_collection.supported_locales, collection_collection.website, collection_collection.description, collection_collection.external_id, collection_collection.custom_validation_schema, collection_collection.references, collection_collection.collection_type, collection_collection.preferred_source, collection_collection.repository_type, collection_collection.custom_resources_linked_source FROM collection_collection WHERE (collection_collection.is_active = True  AND collection_collection.parent_id = 5d721432394d010a2a73c0e3  AND collection_collection.parent_type_id = 5d664271394d01003b4a8133  AND ((collection_collection.parent_id = 5d721432394d010a2a73c0e3  AND collection_collection.parent_type_id = 5d664271394d01003b4a8133 ) OR (collection_collection.parent_type_id = 5d664271394d01003b4a8131  AND collection_collection.parent_id IN (5d721432394d010a2a73c0e8)))) [<Collection: 86285>, <Collection: 86772>]
{
        "op" : "query",
        "ns" : "ocl.collection_collection",
        "query" : {
                "find" : "collection_collection",
                "filter" : {
                        "$or" : [
                                {
                                        "parent_id" : "5d721432394d010a2a73c0e3",
                                        "parent_type_id" : ObjectId("5d664271394d01003b4a8133")
                                },
                                {
                                        "parent_id" : {
                                                "$in" : [
                                                        "5d721432394d010a2a73c0e8"
                                                ]
                                        },
                                        "parent_type_id" : ObjectId("5d664271394d01003b4a8131")
                                }
                        ],
                        "parent_id" : "5d721432394d010a2a73c0e3",
                        "is_active" : true,
                        "parent_type_id" : ObjectId("5d664271394d01003b4a8133")
                },
                "ntoreturn" : 21
        },
        "ts" : ISODate("2019-09-06T08:09:23.173Z")
}



qs from both
SELECT collection_collection.id, collection_collection.public_access, collection_collection.created_at, collection_collection.updated_at, collection_collection.created_by, collection_collection.updated_by, collection_collection.is_active, collection_collection.extras, collection_collection.uri, collection_collection.mnemonic, collection_collection.parent_type_id, collection_collection.parent_id, collection_collection.name, collection_collection.full_name, collection_collection.default_locale, collection_collection.supported_locales, collection_collection.website, collection_collection.description, collection_collection.external_id, collection_collection.custom_validation_schema, collection_collection.references, collection_collection.collection_type, collection_collection.preferred_source, collection_collection.repository_type, collection_collection.custom_resources_linked_source FROM collection_collection WHERE ((collection_collection.is_active = True  AND collection_collection.parent_id = 5d721432394d010a2a73c0e3  AND collection_collection.parent_type_id = 5d664271394d01003b4a8133  AND NOT (collection_collection.public_access = None )) OR (collection_collection.is_active = True  AND collection_collection.parent_id = 5d721432394d010a2a73c0e3  AND collection_collection.parent_type_id = 5d664271394d01003b4a8133  AND ((collection_collection.parent_id = 5d721432394d010a2a73c0e3  AND collection_collection.parent_type_id = 5d664271394d01003b4a8133 ) OR (collection_collection.parent_type_id = 5d664271394d01003b4a8131  AND collection_collection.parent_id IN (5d721432394d010a2a73c0e8))))) [<Collection: 86285>, <Collection: 3870>, <Collection: 86772>, <Collection: 53710>]
{
        "op" : "query",
        "ns" : "ocl.collection_collection",
        "query" : {
                "find" : "collection_collection",
                "filter" : {
                        "$or" : [
                                {
                                        "parent_id" : "5d721432394d010a2a73c0e3",
                                        "public_access" : {
                                                "$ne" : "None"
                                        },
                                        "is_active" : true,
                                        "parent_type_id" : ObjectId("5d664271394d01003b4a8133")
                                },
                                {
                                        "parent_id" : "5d721432394d010a2a73c0e3",
                                        "parent_type_id" : ObjectId("5d664271394d01003b4a8133")
                                },
                                {
                                        "parent_id" : {
                                                "$in" : [
                                                        "5d721432394d010a2a73c0e8"
                                                ]
                                        },
                                        "parent_type_id" : ObjectId("5d664271394d01003b4a8131")
                                },
                                {
                                        "parent_id" : "5d721432394d010a2a73c0e3",
                                        "is_active" : true,
                                        "parent_type_id" : ObjectId("5d664271394d01003b4a8133")
                                }
                        ],
                },
                "ntoreturn" : 21
        },
        "ts" : ISODate("2019-09-06T08:09:23.178Z")
}
```

**Comment by rkorytkowski** [comment]
Thanks @Karuhanga for further investigation. I'll be looking into that on Monday.

**Comment by Karuhanga** [comment]
@rkorytkowski so I managed to find a work around for now.
I have forked `mongo-db-engine` and added tests to replicate this(https://github.com/Karuhanga/mongodb-engine/commit/77550147fe1ed657c99fcac22ec13f6348966907).
Feel free to check these out if you ever get the bandwidth to fix it. I'll try to look into it, but I am prioritising issues on the repo.

**Comment by Karuhanga** [comment]
Awaiting resolution of https://github.com/OpenConceptLab/ocl_issues/issues/151.

**Comment by rkorytkowski** [comment]
I think I might have coincidentally fixed the issue in commits I've just pushed. Could you please verify?

**Comment by Karuhanga** [comment]
> I think I might have coincidentally fixed the issue in commits I've just pushed. Could you please verify with the latest master?

It appears so. However it seems like a new one has cropped up.
cf:
https://api.qa.openconceptlab.org/collections/?q=Roya*&limit=10&page=1&verbose=true
and
https://api.qa.openconceptlab.org/collections/?q=Roya*&limit=10&page=1&verbose=true&user=admin.
The one with the user query does not respect the `q` param.
Also, I am worried about the duplication in the source and collection tests going forward.

**Comment by rkorytkowski** [comment]
@Karuhanga, thanks for testing. I'll look into the bug you found. I refactored tests as time allowed to eliminate duplication. At some point one has to simply stop refactoring ;) You are always welcome to refactor further.

**Comment by rkorytkowski** [comment]
Fixed user param query on collections in https://github.com/OpenConceptLab/oclapi/commit/ec1734c1dbfeb0ad7e83a0f708f2cfb2bed2366c

**Pull request closed without merging**